### DISCUSSION TITLE
[Feature Request] Add a new log file for charts

### DISCUSSION THREAD

**Discussion by natano** [TeamRizu/OutFox#265] [03 - Feature Requests]
**Please give System Info**

OS: Linux Mint 20

CPU: Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz

GPU: Intel Corporation HD Graphics 620 (rev 02)

Storage: Some SSD

**Please give your StepMania Info**

StepMania Version: 5.3 alpha 4.9.1
Theme: Simply Love ([wip-outfox branch](https://github.com/Simply-Love/Simply-Love-SM5/tree/wip-outfox))

**Give an example of what is wrong**

There are many charts out there. Some of them are broken in one way or another. I would expect SM to display a warning when it encounters a broken chart. At the moment this doesn't seem to be the case, at least for the issues that I found.

Here are some examples:
[Chaos in Progress/Tragedy/Tragedy.sm](https://github.com/TeamRizu/OutFox/files/5553200/Tragedy.sm.txt): DISPLAYBPM contains '=' instead of ':'. The song is bpm is correctly displayed as 124 bpm though.
[Cirque du Beast/[Serenade] - 30MIN HARDER/04 - 30MIN HARDER.sm](https://github.com/TeamRizu/OutFox/files/5553192/04.-.30MIN.HARDER.sm.txt): Missing semicolon after the BPMS value.The song does play at correctly 171 bpm though.
[Easy As Pie 4/Begging/Begging.sm](https://github.com/TeamRizu/OutFox/files/5553194/Begging.sm.txt): Backspace character in the BPMS list.
[joe#2 best of collection #5/Vancouver Beatdown/vancouver.sm](https://github.com/TeamRizu/OutFox/files/5553197/vancouver.sm.txt): DISPLAYBPM contains '=' instead of ':'. The song is bpm is correctly displayed as 128 bpm though.
[Mute Sims 3/Beach Blast/beachblast.sm](https://github.com/TeamRizu/OutFox/files/5553193/beachblast.sm.txt): The DISPLAYBPM is set to '?'. I think '*' was meant instead. The bpm displays as 80-151. The song is bpm is correctly displayed as 128 bpm though.
[SHARPNELSTREAMZ v2/I'm A Maid (C-Type Remix)/Im a maid.sm](https://github.com/TeamRizu/OutFox/files/5553195/Im.a.maid.sm.txt): There is a semicolon instead of a colon after the cart description in NOTES. The song is silently ignored and not shown in the song list.
[The Legend of Zim 3/(ZIM) - Prelude/Prelude.sm](https://github.com/TeamRizu/OutFox/files/5553196/Prelude.sm.txt): There is a colon instead of a semicolon after the DISPLAYBPM value. The bpm is correctly displayed as 140 bpm though.

**Given an example of what it should look like**

I would expect a warning to show in the SM output when a chart has an error, similar to how lua errors are displayed.

I think step artists would be more likely to notice errors in their stepcharts if a warning would be to displayed in the SM output.

Logs: [Logs.tar.gz](https://github.com/TeamRizu/OutFox/files/5553254/Logs.tar.gz)

**Comment by Scraticus** [comment]
The issue with this request, is the log already is spammed with errors and such in simfiles. The only way to make this 'clean' would be to have a new logfile just for simfiles, so we aren't having to go through hundreds of lines, as many of the bugs are just not fixed in sim errors. 

I'd suggest we either use 'info.txt' like we do for song issues (files not opening) and allow the main log to be a bit cleaner for debugging.

Thoughts?

**Comment by natano** [comment]
The idea was to have the errors in-your-face, so that step artists actually notice the errors when test-playing their charts, so they can fix them before releasing. But yes, that might inconvenience users that have a broken chart they downloaded from somewhere. I don't have a good solution for that.

If simfile parse errors would show up on stdout/stderr like Warn()s I would probably notice them, but I'm not sure how many stepartists run SM from the terminal. On the other hand I don't think many people would proactively check a log file for errors in their charts either.

Of course having errors in the info log (or even a new file) would be an improvement over the current situation of having no indication of an error at all.

**Comment by Scraticus** [comment]
Ah, we are planning to revamp the debug overlay screen. F2 will be
mod/stepcharter specific (and how many times have you caught it by accident
and reloaded everything?!)

As we move to a more mature system, I'd like a better method for displaying
the last errors and such in the debug overlay, and I always felt an F2
overlay would be perfect for this.

As for legacy/ancient charts, we could have a "warn on chart errors"
preference as the average person isn't really gonna bother trying to fix a
chart made in 2007 for example.

Some interesting points. What else would be useful in a new debug overlay?

**Comment by natano** [comment]
Ah, I didn't know about those plans. Sounds good to me!

Regarding debugging: It would  actually be amazing to have an interactive lua shell available in SM that allows to inspect variables or run short snippets.
